Firestore/core/test/firebase/firestore/local/counting_query_engine.h

```diff
@@ -0,0 +1,204 @@
+    return documents_read_by_key_;
+  }
+
+  /**
+   * Returns the number of mutations returned by the MutationQueue's
+   * `getAllMutationBatchesAffectingQuery()` API (since the last call to
+   * `ResetCounts()`)
+   */
+  int mutations_read_by_query() {
+    return mutations_read_by_query_;
+  }
+
+  /**
+   * Returns the number of mutations returned by the MutationQueue's
+   * `AllMutationBatchesAffectingDocumentKey()` and
+   * `AllMutationBatchesAffectingDocumentKeys()` APIs (since the last call to
+   * `ResetCounts()`)
+   */
+  int mutations_read_by_key() {
+    return mutations_read_by_key_;
+  }
+
+ private:
+  friend class WrappedMutationQueue;
+  friend class WrappedRemoteDocumentCache;
+
+  QueryEngine* query_engine_;
+  bool index_free_;
+
+  std::unique_ptr<LocalDocumentsView> local_documents_;
+  std::unique_ptr<WrappedMutationQueue> mutation_queue_;
+  std::unique_ptr<WrappedRemoteDocumentCache> remote_documents_;
+
+  int mutations_read_by_query_ = 0;
+  int mutations_read_by_key_ = 0;
+  int documents_read_by_query_ = 0;
+  int documents_read_by_key_ = 0;
+};
+
+/** A MutationQueue that counts document reads. */
+class WrappedMutationQueue : public MutationQueue {
+ public:
+  WrappedMutationQueue(MutationQueue* subject,
+                       CountingQueryEngine* query_engine)
+      : subject_(subject), query_engine_(query_engine) {
+  }
+
+  void Start() override;
+
+  bool IsEmpty() override;
+
+  void AcknowledgeBatch(const model::MutationBatch& batch,
+                        const nanopb::ByteString& stream_token) override;
+
+  model::MutationBatch AddMutationBatch(
+      const Timestamp& local_write_time,
+      std::vector<model::Mutation>&& base_mutations,
+      std::vector<model::Mutation>&& mutations) override;
+
+  void RemoveMutationBatch(const model::MutationBatch& batch) override;
+
+  std::vector<model::MutationBatch> AllMutationBatches() override;
+
+  std::vector<model::MutationBatch> AllMutationBatchesAffectingDocumentKeys(
+      const model::DocumentKeySet& document_keys) override;
+
+  std::vector<model::MutationBatch> AllMutationBatchesAffectingDocumentKey(
+      const model::DocumentKey& key) override;
+
+  std::vector<model::MutationBatch> AllMutationBatchesAffectingQuery(
+      const core::Query& query) override;
+
+  absl::optional<model::MutationBatch> LookupMutationBatch(
+      model::BatchId batch_id) override;
+
+  absl::optional<model::MutationBatch> NextMutationBatchAfterBatchId(
+      model::BatchId batch_id) override;
+
+  model::BatchId GetHighestUnacknowledgedBatchId() override;
+
+  void PerformConsistencyCheck() override;
+
+  nanopb::ByteString GetLastStreamToken() override;
+
+  void SetLastStreamToken(nanopb::ByteString stream_token) override;
+
+ private:
+  MutationQueue* subject_;
+  CountingQueryEngine* query_engine_;
+};
+
+/** A RemoteDocumentCache that counts document reads. */
+class WrappedRemoteDocumentCache : public RemoteDocumentCache {
```

Comment:
Port of https://github.com/firebase/firebase-android-sdk/blob/2a2e789629a7dad13e1007b6cdec26c96e73defe/firebase-firestore/src/test/java/com/google/firebase/firestore/local/CountingQueryEngine.java#L116